File: leet-code/1. Two Sum/1_two_sum.py
def two_sum(nums, target):
    dic = dict()

    for idx, num in enumerate(nums):
        dic[num] = idx

    for idx, num in enumerate(nums):
        # if target - num in nums and idx != dic[target - num]:
        # 위처럼 list 에 target - num 이 있는지 조회하는 것 보다
        # 아래처럼 dictionary 에 target - num 이 있는지 조회하는게 훨씬 빠르다
        if target - num in dic and idx != dic[target - num]:
            return [idx, dic[target - num]]


print(two_sum([2, 7, 11, 15], 9))

File: leet-code/1. Two Sum/1_two_sum.dart
class Solution {
  List<int> twoSum(List<int> nums, int target) {
    Map<int, int> numMap = Map<int, int>();

    for (int i = 0; i < nums.length; i++) {
      numMap[nums[i]] = i;
    }

    for (int i = 0; i < nums.length; i++) {
      if (numMap.containsKey(target - nums[i]) && i != numMap[target - nums[i]]) {
        return [i, numMap[target - nums[i]]!];
      }
    }
    
    return [];
  }
}

void main() {
  Solution solution = Solution();
  print(solution.twoSum([2, 7, 11, 15], 9));
  print(solution.twoSum([3, 2, 4], 6));
}
